File: pkg/util/semver.go
package util

import (
	"errors"
	"regexp"
	"strconv"
	"strings"

	"github.com/factrylabs/factry-historian-datasource.git/pkg/schemas"
)

// SemVer represents a semantic version
type SemVer struct {
	Major         int
	Minor         int
	Patch         int
	PreRelease    string
	BuildMetadata string
	IsDebug       bool
}

var semVerRegex = regexp.MustCompile(`^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|[0-9a-zA-Z-]+)(?:\.(?:0|[1-9]\d*|[0-9a-zA-Z-]+))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`)

func parseSemVer(version string) (SemVer, error) {
	if version == "debug" {
		return SemVer{
			Major:   int(^uint(0) >> 1), // Maximum integer value
			Minor:   int(^uint(0) >> 1),
			Patch:   int(^uint(0) >> 1),
			IsDebug: true,
		}, nil
	}

	matches := semVerRegex.FindStringSubmatch(version)
	if matches == nil {
		return SemVer{}, errors.New("invalid semantic version string")
	}

	major, _ := strconv.Atoi(matches[1])
	minor, _ := strconv.Atoi(matches[2])
	patch, _ := strconv.Atoi(matches[3])

	return SemVer{
		Major:         major,
		Minor:         minor,
		Patch:         patch,
		PreRelease:    matches[4],
		BuildMetadata: matches[5],
		IsDebug:       false,
	}, nil
}

func toStringSemVer(semVer SemVer) string {
	if semVer.IsDebug {
		return "debug"
	}

	version := strconv.Itoa(semVer.Major) + "." + strconv.Itoa(semVer.Minor) + "." + strconv.Itoa(semVer.Patch)

	if semVer.PreRelease != "" {
		version += "-" + semVer.PreRelease
	}

	if semVer.BuildMetadata != "" {
		version += "+" + semVer.BuildMetadata
	}

	return version
}

func comparePreRelease(a, b string) int {
	aParts := strings.Split(a, ".")
	bParts := strings.Split(b, ".")

	maxLen := len(aParts)
	if len(bParts) > maxLen {
		maxLen = len(bParts)
	}

	for i := 0; i < maxLen; i++ {
		var aPart, bPart string
		if i < len(aParts) {
			aPart = aParts[i]
		}
		if i < len(bParts) {
			bPart = bParts[i]
		}

		aIsNum := isNumeric(aPart)
		bIsNum := isNumeric(bPart)

		switch {
		case aIsNum && bIsNum:
			aNum, _ := strconv.Atoi(aPart)
			bNum, _ := strconv.Atoi(bPart)
			if aNum != bNum {
				return aNum - bNum
			}
		case aIsNum:
			return -1
		case bIsNum:
			return 1
		case aPart != bPart:
			if aPart < bPart {
				return -1
			}
			return 1
		}
	}

	return 0
}

// SemverCompare compares two semantic versions
func SemverCompare(a, b string) int {
	aSemVer, err := parseSemVer(a)
	if err != nil {
		return 1
	}

	bSemVer, err := parseSemVer(b)
	if err != nil {
		return -1
	}

	switch {
	case aSemVer.IsDebug && bSemVer.IsDebug:
		return 0
	case aSemVer.IsDebug:
		return 1
	case bSemVer.IsDebug:
		return -1
	case aSemVer.Major != bSemVer.Major:
		return aSemVer.Major - bSemVer.Major
	case aSemVer.Minor != bSemVer.Minor:
		return aSemVer.Minor - bSemVer.Minor
	case aSemVer.Patch != bSemVer.Patch:
		return aSemVer.Patch - bSemVer.Patch
	case aSemVer.PreRelease != "" && bSemVer.PreRelease != "":
		return comparePreRelease(aSemVer.PreRelease, bSemVer.PreRelease)
	case aSemVer.PreRelease != "":
		return -1
	case bSemVer.PreRelease != "":
		return 1
	}

	return 0
}

func isNumeric(s string) bool {
	_, err := strconv.Atoi(s)
	return err == nil
}

// IsValidSemver checks if a string is a valid semver. Example: 1.2.3-beta.1
func IsValidSemver(version string) bool {
	return semVerRegex.MatchString(version)
}

// CheckMinimumVersion checks if the historian version is not less than minVersion
func CheckMinimumVersion(info *schemas.HistorianInfo, minVersion string, includePreReleases bool) bool {
	if info == nil {
		return false
	}

	historianVersion, _ := strings.CutPrefix(info.Version, "v")

	if includePreReleases {
		semVer, err := parseSemVer(historianVersion)
		if err != nil {
			return true // Assume invalid version is greater than minVersion
		}
		semVer.PreRelease = ""
		historianVersion = toStringSemVer(semVer)
	}

	return SemverCompare(historianVersion, minVersion) >= 0
}

File: src/util/semver.ts
// Define the semVerRegex as a TypeScript regular expression
const semVerRegex =
  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|[0-9a-zA-Z-]+)(?:\.(?:0|[1-9]\d*|[0-9a-zA-Z-]+))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/

// Define types for semantic version components
type SemVer = {
  major: number
  minor: number
  patch: number
  preRelease?: string
  buildMetadata?: string
  isDebug?: boolean
}

// Helper function to parse a semantic version string
function parseSemVer(version: string): SemVer {
  // remove leading 'v' if present
  if (version.startsWith('v')) {
    version = version.slice(1)
  }
  const matches = version.match(semVerRegex)
  if (!matches) {
    return {
      major: Number.MAX_SAFE_INTEGER,
      minor: Number.MAX_SAFE_INTEGER,
      patch: Number.MAX_SAFE_INTEGER,
      isDebug: true,
    }
  }

  return {
    major: parseInt(matches[1], 10),
    minor: parseInt(matches[2], 10),
    patch: parseInt(matches[3], 10),
    preRelease: matches[4],
    buildMetadata: matches[5],
    isDebug: false,
  }
}

// Helper function to convert a SemVer object to a string
function toStringSemVer(semVer: SemVer): string {
  if (semVer.isDebug) {
    return 'debug'
  }

  let version = `${semVer.major}.${semVer.minor}.${semVer.patch}`

  if (semVer.preRelease) {
    version += `-${semVer.preRelease}`
  }

  if (semVer.buildMetadata) {
    version += `+${semVer.buildMetadata}`
  }

  return version
}

// Helper function to compare two pre-release components
function comparePreRelease(a: string, b: string): number {
  const aParts = a.split('.')
  const bParts = b.split('.')

  for (let i = 0; i < Math.max(aParts.length, bParts.length); i++) {
    const aPart = aParts[i] || ''
    const bPart = bParts[i] || ''

    // Compare numeric parts
    const aIsNum = /^\d+$/.test(aPart)
    const bIsNum = /^\d+$/.test(bPart)

    if (aIsNum && bIsNum) {
      const diff = parseInt(aPart, 10) - parseInt(bPart, 10)
      if (diff !== 0) {
        return diff
      }
    } else if (aIsNum) {
      return -1 // Numeric identifiers are less than non-numeric
    } else if (bIsNum) {
      return 1
    } else {
      // Lexicographical comparison for strings
      if (aPart !== bPart) {
        return aPart < bPart ? -1 : 1
      }
    }
  }

  return 0
}

// Helper function to compare two semantic version strings
export function semverCompare(a: string, b: string): number {
  const aSemVer = parseSemVer(a)
  const bSemVer = parseSemVer(b)

  // Check for debug versions
  if (aSemVer.isDebug && bSemVer.isDebug) {
    return 0
  } else if (aSemVer.isDebug) {
    return 1
  } else if (bSemVer.isDebug) {
    return -1
  }

  // Compare the major, minor, and patch versions
  if (aSemVer.major !== bSemVer.major) {
    return aSemVer.major - bSemVer.major
  }

  if (aSemVer.minor !== bSemVer.minor) {
    return aSemVer.minor - bSemVer.minor
  }

  if (aSemVer.patch !== bSemVer.patch) {
    return aSemVer.patch - bSemVer.patch
  }

  // Compare the pre-release versions if both exist
  if (aSemVer.preRelease && bSemVer.preRelease) {
    return comparePreRelease(aSemVer.preRelease, bSemVer.preRelease)
  } else if (aSemVer.preRelease) {
    return -1 // Pre-release versions are considered less than normal versions
  } else if (bSemVer.preRelease) {
    return 1
  }

  // Ignore build metadata; versions are equal
  return 0
}

// Helper function to check if a version is greater than or equal to a target version, excluding pre-releases by default
export function isFeatureEnabled(version: string, targetVersion: string, includePreReleases = false): boolean {
  if (includePreReleases) {
    let versionSemVer = parseSemVer(version)
    versionSemVer.preRelease = undefined
    version = toStringSemVer(versionSemVer)
  }

  return semverCompare(version, targetVersion) >= 0
}
